### PR TITLE
Add Encoder with Buhlmann-Straub estimator

### PR DESCRIPTION
**Pull request by lorentzenchr**
This adds the `TargetEncoderBS` with Bühlmann-Straub estimation.

Also, shuffling is added to the cv splits and benchmark results are updated.

### CODE REVIEW

**Comment by lorentzenchr** [comment]
image
image
image
image
image
image
image
image

**Comment by lorentzenchr** [comment]
Some datasets deserve special attention:
- telco: As "drop" shows, the categoricals do not contain much information. The splits matter a lot (large variance), which means there are categoricals with either only very few levels or levels that occur very rarely, but have an effect on the target.
- amazon_access: Also here, I suspect rare levels with some effects on the performance.
- dresses_sales: The pipeline with Bühlmann-Straub estimated target encoding seems to outperform all other models, but within uncertainty (variance).
- kicks: The encoding with cv seem to perform worse.

TLDR: I'm not (yet) completely convinced by the CV target encoders...

File: benchmark.py [hunk 1]
```diff
@@ -99,7 +100,9 @@ def run_single_benchmark(data_str, encoder_str, cv, n_jobs, write_result, force)
 
     if data_info.is_classification:
         scoring = ["average_precision", "roc_auc", "accuracy"]
+        cv = StratifiedKFold(n_splits=cv, shuffle=True, random_state=42)
     else:
+        cv = KFold(n_splits=cv, shuffle=True, random_state=42)
```
Comment (thomasjpfan) [inline benchmark.py:105]: For reference: `cross_validate` handles this behind the scenes, when it calls `check_cv`:

https://github.com/scikit-learn/scikit-learn/blob/1e386a49fcaefcc9860266b5957582bc85aa56ab/sklearn/model_selection/_validation.py#L233

**Review by thomasjpfan — COMMENTED** [review]
Thank you for adding the buhlmann-straub encoder!

I am always on the looking out for more datasets to benchmark with. (I may throw in the [datasets from dirty_cat](https://dirty-cat.github.io/stable/#data-download)

I think I want to see what happens with the "Bagging" approach discussed from the PR. This would work around the `fit_transform != `fit.transform` issue.

File: benchmark.py (hunk 1 above)
Comment (lorentzenchr) [reply benchmark.py:105]: One never stops learning.:smile:
